src/index.ts

```diff
@@ -913,6 +934,60 @@ class ThreadPool {
 
     await Promise.all(exitEvents);
   }
+
+  async close (options : CloseOptions = kDefaultCloseOptions) {
+    this.closingUp = true;
+
+    if (options.force) {
+      const skipQueueLength = this.skipQueue.length;
+      for (let i = 0; i < skipQueueLength; i++) {
+        const taskInfo : TaskInfo = this.skipQueue.shift() as TaskInfo;
+        if (taskInfo.workerInfo === null) {
+          taskInfo.done(new AbortError());
+        } else {
+          this.skipQueue.push(taskInfo);
+        }
+      }
+
+      const taskQueueLength = this.taskQueue.size;
+      for (let i = 0; i < taskQueueLength; i++) {
+        const taskInfo : TaskInfo = this.taskQueue.shift() as TaskInfo;
+        if (taskInfo.workerInfo === null) {
+          taskInfo.done(new AbortError());
+        } else {
+          this.taskQueue.push(taskInfo);
+        }
```

Comment:
I'm not fully convinced that aborting them is the right path to choose, I'd prefer to find a way to flush them before a timeout.
My expectation is similar to what happens when you run `httpServer.close`, current requests are being flushed out, but no new ones are being accepted.

Reply:
yeah I wasn't sure to using `AbortError` here, I tried to use a different approach where I flushed the queue, however the `run` promises were fulfilled with `null` (which didn't convinced me, since you expect a task to complete with your expected result or fail in the other case)

if you could point me in the right direction to implement something similar to `httpServer.close` having the above in mind, it would be great

Reply:
This is the rough implementation I was doing:

```ts
  async close (): Promise<void> {
    if (this.closing || this.closed) return;

    this.closing = true;

    // Flush high-prio enqueued tasks
    const result = await Promise.race([
      flushPool.call(this),
      settimeout(this.options.closeTimeout ?? 60000) // default to 1 minute
    ]);

    const err = !result ? new Error('closed timed out') : null;

    await this.destroy();
    this.closed = true;
    this.closing = false;

    if (err) this.publicInterface.emit('error', err);
    this.publicInterface.emit('close', err);

    // TODO: Flush is hanging, it seems it is not getting updates
    // from the workers to account for the number of current tasks in flight
    async function flushPool (this: ThreadPool): Promise<boolean> {
      while (this._getTaskCountInFlight() !== 0) {
        while (this.skipQueue.length !== 0) {
          const taskInfo = <TaskInfo> this.skipQueue.shift();
          const candidate = this.workers.findAvailable();

          if (!candidate || this.needsDrain) {
            await once(this.publicInterface, 'drain');
            this.skipQueue.push(taskInfo);
            continue;
          }

          const now = performance.now();
          this.waitTime.recordValue(now - taskInfo.created);
          taskInfo.started = now;
          candidate.postTask(taskInfo);
          this._maybeDrain();
        }

        // Flush lower prio tasks
        while (this.taskQueue.size !== 0) {
          const taskInfo : TaskInfo = this.taskQueue.shift() as TaskInfo;
          const candidate = this.workers.findAvailable();

          if (!candidate || this.needsDrain) {
            await once(this.publicInterface, 'drain');
            this.taskQueue.push(taskInfo);
            continue;
          }

          const now = performance.now();
          this.waitTime.recordValue(now - taskInfo.created);
          taskInfo.started = now;
          candidate.postTask(taskInfo);
          this._processPendingMessages();
          this._maybeDrain();
        }

	    this._processPendingMessages();
      }
      return true;
    }
  }
```

The important piece is the `flushPool` which is the one seeking to flush all the current and pending tasks within the queue and pool until all.